crates/sui-indexer-alt/src/pipeline/sequential/committer.rs

```diff
@@ -163,6 +163,15 @@ pub(super) fn committer<H: Handler + 'static>(
                         "Gathered batch",
                     );
 
+                    if batch_checkpoints == 0 {
```

Comment:
I don't think we can skip the entire rest of the body, because at the end of the tick, we check whether we've processed everything and the `rx` channel was closed which is our condition for gracefully shutting down the pipeline (no more data is coming in, and we've processed all the data we can). So I think with this change, the pipeline will just hang once it is done processing checkpoints (if you set a `--last-checkpoint`).

IIUC, the main thing you want to avoid doing is interacting with the DB in these scenarios, which I think you can do most directly with a change like this, around where we currently create the DB transaction:

```rust
let affected = if batch_rows == 0 && !watermark_needs_update {
    Ok(0)
} else {
    let Ok(mut conn) = db.connect().await else {
        warn!(pipeline = H::NAME, "Failed to get connection for DB");
        continue;
    };

    conn.transaction::<_, anyhow::Error, _>(|conn| async {
        watermark.update(conn).await?;
        H::commit(&batch, conn).await;
    }.scope_boxed()).await
}
```

This means we still get to hit the termination condition for the task at the end, and it's clear why we can avoid the transaction if `batch_rows == 0 && !watermark_needs_update`.

I've also pulled the connection into the `if` which means we avoid taxing the connection pool unnecessarily, and we also don't need to drop the connection early as well, so overall this seems like a worthwhile change.

Reply:
There are a few things I would like to improve as well:
- We might want to avoid pollute some metrics when batch size is 0.
- It doesn't seem that we actually need `watermark_needs_update`. It is always equal to `batch_checkpoints > 0`.
- We might want to consider combining the conditions that would exit the loop to make it clearer. Right now we deal with lag exit condition at the beginning, and general condition at the end.

Reply:
The first two points sound good to me. I'm not sure I understand the last bullet though (mainly, I don't know what the "lag exit condition" is -- I don't think we have that, unless I've missed something).

Reply:
Updated PR: consolidated the exit condition so that it is clear on when we would break the loop.